test/integration/tcp_proxy_integration_test.cc

```diff
@@ -34,9 +34,10 @@ TEST_P(TcpProxyIntegrationTest, TcpProxyUpstreamWritesFirst) {
   tcp_client->write("hello");
   fake_upstream_connection->waitForData(5);
 
-  fake_upstream_connection->close();
+  fake_upstream_connection->halfClose();
```

Comment:
Now that we have half close, can we modify these so we 
half close upstream->downstream path
wait for half close to propogate
write on downstream->upstream path (verify the connection is still open)
fully close
and vice versa?

Reply:
Yep, it's on my TODO list.  There's a whole bunch of tests that need to be written for this PR still.